modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/GlobalClassAnalyzer.cs

```diff
@@ -0,0 +1,103 @@
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Godot.SourceGenerators
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class GlobalClassAnalyzer : DiagnosticAnalyzer
+    {
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => ImmutableArray.Create(
+                Common.GlobalClassMustImplementGodotObjectRule,
+                Common.GlobalClassMustNotBeGenericRule,
+                Common.ParentClassMustBeGlobalRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.ClassDeclaration);
+        }
+
+        private void AnalyzeNode(SyntaxNodeAnalysisContext context)
+        {
+            // Ignore syntax inside comments
+            if (IsInsideDocumentation(context.Node))
+                return;
+
+            var typeClassDecl = (ClassDeclarationSyntax)context.Node;
+
+            if (context.ContainingSymbol is not INamedTypeSymbol typeSymbol)
+                return;
+
+            // Search attributes for [GlobalClass] attribute
+            var attrs = typeSymbol?.GetAttributes();
+            if (attrs == null)
+                return;
+
+            bool isGlobal = false;
+            foreach (var attribute in attrs)
+            {
+                if (!(attribute.AttributeClass?.IsGodotGlobalClassAttribute() ?? false))
+                    continue;
+
+                isGlobal = true;
+                break;
+            }
+
+            if (!isGlobal)
+                return;
+
+            if (typeSymbol?.IsGenericType ?? false)
+                Common.ReportGlobalClassMustNotBeGeneric(context, typeClassDecl, typeSymbol);
+
+            if (!typeSymbol.InheritsFrom("GodotSharp", GodotClasses.GodotObject))
+            {
+                Common.ReportGlobalClassMustImplementGodotObject(context, typeClassDecl, typeSymbol!);
+                // Return since there is a chance that there is no parent.
+                return;
+            }
+
+            // I'd prefer if the GodotSharp stuff was already pre-annotated with [GlobalClass]...
+            // Suppresses a stupid 'editor\create_dialog.cpp:266 - Condition "inherits.is_empty()" is true.'
+            var baseType = typeSymbol!.BaseType;
+            if (!IsObjectFromGodotSharp(baseType) && !(baseType?.GetAttributes()
+                .Any(x => x.AttributeClass?.IsGodotGlobalClassAttribute() ?? false) ?? false))
+                Common.ReportParentClassMustBeGlobal(context, typeClassDecl, typeSymbol);
```

Comment:
Ah, I hadn't tested this in GDScript yet. It seems there, though, that the GDScript parent still shows up in the editor (unselectable, but it _is_ there):

![image](https://github.com/godotengine/godot/assets/46660281/937ac3dd-76c9-4817-b76a-5d147ae64f16)

However, the C# parent does not. This is related to the weird error commented above, where the create dialog cannot find the C# parent type and throws an error. Unfortunately, that is out of the scope for this PR.

I'll remove the check and related diagnostic since it seems a little unnecessary now.